Track base of FunctionalTensor in inference mode.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135284
* __->__ #135141

The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) then we consider it is as the base of the all the tensors that shares its storage.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

laithsakka committed:
support inference mode

[ghstack-poisoned]

laithsakka commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -158,6 +161,15 @@ def __new__(cls, elem):
+                out._inference_mode_base = None
+                mode._storage_to_base[out.elem.storage()._cdata] = out
+            else:
+                out._inference_mode_base = mode._storage_to_base[
```
is there cases where we wont see mode._storage_to_base does not exists? shall we be safe here and assign it to none if not found?

zou3519 replied:
I don't think so? We'd prefer an error

zou3519 commented on aten/src/ATen/FunctionalTensorWrapper.h:
```diff
@@ -165,6 +165,11 @@ struct TORCH_API FunctionalTensorWrapper : public c10::TensorImpl {
     was_storage_changed_ = true;
   }
 
+  // A tensor is a root tensor, if its not a view of another tensor.
+  bool isRootTensor() const {
+    return view_metas_.empty();
+  }
```
nit: naming: maybe just call this "Base"? isBaseTensor()

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -158,6 +161,15 @@ def __new__(cls, elem):
+        if torch.is_inference_mode_enabled():
+            if out.isRootTensor():
+                out._inference_mode_base = None
+                mode._storage_to_base[out.elem.storage()._cdata] = out
```
you probably want to store a weakref.ref to the out.elem.storage(). It's possible that the tensor can go out of scope and the cdata can be reused.

zou3519 replied:
Or use a weakref.WeakKeyDictionary. The key should be the out.elem.storage()

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -226,14 +239,17 @@ def to_functional(x):
         torch._sync(self)
         return torch._from_functional_tensor(self.elem)
 
+    def isRootTensor(self) -> bool:
```
nit: we usually do snake case in python

laithsakka replied:
ah yes ops

laithsakka replied:
I guess i followed isFunctionalTensor

zou3519 reviewed (commented):
code looks good to me, I want to bikeshed the naming

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -316,6 +336,10 @@ def __init__(self, pre_dispatch=False, export=False, _allow_token_discovery=Fals
         self._allow_token_discovery = _allow_token_discovery
 
+        self._storage_to_base: Dict[
+            torch.storage.UntypedStorage, Optional[weakref.ReferenceType[FunctionalTensor]]
```
As discussed offline we probably want WeakKeyDictionary[UntypedStorage, FunctionalTensor]

zou3519 commented on torch/csrc/autograd/python_torch_functions_manual.cpp:
```diff
@@ -664,6 +664,10 @@ void initTorchFunctions(PyObject* module) {
             !at::functionalization::impl::isFunctionalTensor(o));
         at::functionalization::impl::replace_(t, o);
       });
+  py_module.def("_is_functional_tensor_root", [](const at::Tensor& t) {
```
root -> base

zou3519 commented on aten/src/ATen/FunctionalTensorWrapper.h:
```diff
@@ -165,6 +165,11 @@ struct TORCH_API FunctionalTensorWrapper : public c10::TensorImpl {
     was_storage_changed_ = true;
   }
 
+  // A tensor is a considered base, if its not a view of another tensor.
```
nit: "A FunctionalTensor is considered a base if..."

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -158,6 +162,18 @@ def __new__(cls, elem):
+        if torch.is_inference_mode_enabled():
+            if out.is_base_tensor():
+                out._inference_mode_base = None
+                # This assumes that the FunctionalTensor does not change its storage after this point.
```
Concretely, we are assuming that `out.elem` does not change its storage.

zou3519 replied:
I mention this because we do change out's storage after the `__new__` (on view operations, we patch the out tensor to have the same storage as the input to the view)

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Track base of FunctionalTensor in inference mode. "


The idea behind the tracking is the following, whenever we see a tensor if the tensors is a root tensors (does not have any view metas ) when we consider is as the base of the all the tensors that shares its storage.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka commented:
rebase on main

laithsakka commented:
@pytorchbot merge

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -158,6 +162,21 @@ def __new__(cls, elem):
+            torch.is_inference_mode_enabled()
+            and torch._inductor.config.enable_auto_functionalized_v2
+        ):
+            if out.is_base_tensor():
```
mentioned to @zou3519 offline, but I think it would be nice if we could kill this logic in favor of doing it directly in the C++ `FunctionalTensorWrapper`. in theory, we have all of the state that we need:

(1) current `FunctionalTensorWrapper` (view) tensor
(2) its `FunctionalStorageImpl::value`, aka the underlying (`FakeTensor`) base tensor

and performing the update for `updated_base, updated_view = auto_functionalize(mutable_op, base, view)` would look something like:

(1) sync the view to flush any pending mutations first
(2) perform `view_updated = mutable_op(clone(view)))`
(3) get the updated base with `updated_base = view_inverses(view_updated)`
(4) manually update the (shared) `FunctionalStorageImpl`, so its inner value gets replaced with `updated_base`, and update the counter on the storage so other aliases know the base was mutated

I think this should be mostly doable with the existing API's (e.g. `apply_updates`: https://github.com/pytorch/pytorch/blob/04118d86172e674865e9b6a12c1fbaa075fa7b00/aten/src/ATen/FunctionalStorageImpl.cpp#L132), although we might need to make a few tweaks

zou3519 replied:
@bdhirsh We tried some number of similar designs and ran into problems with all of them, but it's possible that we didn't find the right design. The problem with this particular approach is that `updated_base = view_inverses(view_updated)` emits a number of view_scatter operations into the graph that make it difficult to do the reinplacing later on (the reinplacing would need to know it can kill those view_scatter ops). The approach we've taken does not end up emitting view_scatter operations and it does this by doing
- Emit base_updated = auto_functionalized(mutable_op, base, view) into the graph (thereby hiding the scatters inside of auto_functionalized)
- replace/commit_update/sync on (base, base_updated)
- any views of base that get used in the future now are based on the `base_updated` Tensor. Using them will reconstruct the views, emitting view ops into the graph. These view ops don't need to be cleaned up later.
